lib/src/gzip_encoding.dart

```diff
@@ -0,0 +1,230 @@
+  var bytesBuffer = await response.read().fold<_BytesBuffer>(
+      _BytesBuffer(bufferInitialCapacity),
+      (result, bytes) => result..addAll(bytes));
+
+  var gzipEncoder = compressionLevel == defaultGzipEncodingCompressionLevel
+      ? _defaultGzipEncoder
+      : ZLibEncoder(gzip: true, level: compressionLevel);
+
+  // Compressed body:
+  var compressedBody = gzipEncoder.convert(bytesBuffer.toUint8List());
+
+  var bodyLength = bytesBuffer.length;
+  var compressedBodyLength = compressedBody.length;
+  var compressionRatio = compressedBodyLength / bodyLength;
+
+  var headers = Map<String, String>.from(response.headers);
+
+  headers[HttpHeaders.contentEncodingHeader] = 'gzip';
+  headers[HttpHeaders.contentLengthHeader] = compressedBodyLength.toString();
+  headers['X-Compression-Ratio'] =
+      '$compressionRatio ($compressedBodyLength/$bodyLength)';
+
+  return response.change(headers: headers, body: compressedBody);
+}
+
+/// Returns `true` if [response] can be compressed.
+///
+/// Checks:
+/// - `Content-Encoding`: if already present, can't change it to `gzip`.
+/// - `Content-Type`: checks if [isAlreadyCompressedContentType].
+/// - `Content-Length`: checks if too small for compression (< 512).
+bool canGzipEncodeResponse(Response response) {
+  var headerContentEncoding =
+      response.headers[HttpHeaders.contentEncodingHeader];
+
+  // If the response already defines a `Content-Encoding` header it
+  // won't apply the the `gzip` encoding to preserve the response behavior.
+  if (headerContentEncoding != null && headerContentEncoding.isNotEmpty) {
+    return false;
+  }
+
+  var responseContentLength = response.contentLength;
+
+  // A small body will not benefit from being compressed:
+  if (responseContentLength != null && responseContentLength < 512) {
+    return false;
+  }
+
+  var headerContentType = response.headers[HttpHeaders.contentTypeHeader];
+
+  // Do not compress if the `Content-Type` is an already compressed type:
+  if (headerContentType != null &&
+      isAlreadyCompressedContentType(headerContentType)) {
+    return false;
+  }
+
+  return true;
+}
+
+/// Returns `true` if [contentType] is already compressed.
+///
+/// See [isAlreadyCompressedExtension].
+bool isAlreadyCompressedContentType(String contentType) {
+  contentType = contentType.toLowerCase().trim();
+  if (contentType.isEmpty) return false;
+
+  var idx = contentType.indexOf(';');
+  if (idx >= 0) {
+    contentType = contentType.substring(0, idx).trim();
+  }
+
+  var idx2 = contentType.indexOf('/');
+
+  var type = idx2 >= 0 ? contentType.substring(idx2 + 1) : contentType;
+
+  if (isAlreadyCompressedExtension(type)) {
+    return true;
+  }
+
+  if (type.contains('+')) {
+    var list = type.split('+');
+    var alreadyCompressed =
+        list.where((e) => isAlreadyCompressedExtension(e)).isNotEmpty;
+    return alreadyCompressed;
+  }
+
+  return false;
+}
+
+bool isAlreadyCompressedExtension(String extension) {
```

Comment:
Where does this list come from? It might be better to allow users to provide their own list of types to exclude.

Reply:
This is a list of common extensions that are always compressed.

Reply:
To allow personalization the structure should be changed to a more Object Oriented pattern. I tried to respect the previous pattern of the project.

Reply:
I'm not sure what you mean - why not allow this list to be overridden?

Reply:
I have updated the code and now the developer can pass a parameter `alreadyCompressedContentType`, where it can overwrite the function that determines if a content is already compressed.